### PR TITLE
feat/New functions to blockchain hook

### PR DESCRIPTION
**Pull request by sasurobert**
New functions to blockchain hook.

### CODE REVIEW

File: process/block/metablock.go [hunk 1]
```diff
@@ -64,6 +64,7 @@ func NewMetaProcessor(arguments ArgMetaProcessor) (*metaProcessor, error) {
 		onRequestHeaderHandler:        arguments.RequestHandler.RequestHeader,
 		onRequestHeaderHandlerByNonce: arguments.RequestHandler.RequestHeaderByNonce,
 		appStatusHandler:              statusHandler.NewNilStatusHandler(),
+		blockChainHook:                arguments.BlockChainHook,
```
Comment (iulianpascalau) [inline process/block/metablock.go]: Should check this for nil

File: process/smartContract/hooks/vmAccountsDB.go [hunk 2]
```diff
@@ -141,9 +181,106 @@ func (vadb *VMAccountsDB) GetCode(address []byte) ([]byte, error) {
...
+
+// LastEpoch returns the epoch from the last committed block
+func (vadb *VMAccountsDB) LastEpoch() uint32 {
+	return vadb.blockChain.GetCurrentBlockHeader().GetEpoch()
+}
+
+// GetStateRootHash returns the state root hash from the last committed block
+func (vadb *VMAccountsDB) GetStateRootHash() []byte {
+	return vadb.blockChain.GetCurrentBlockHeader().GetRootHash()
+}
```
Comment (iulianpascalau) [inline process/smartContract/hooks/vmAccountsDB.go]: Should add a check for nil
```GO
if adb.blockChain.GetCurrentBlockHeader() != nil{
     //.....
}
```

File: consensus/spos/commonSubround/subroundBlock.go [hunk 3]
```diff
@@ -186,22 +193,8 @@ func (sr *SubroundBlock) sendBlockHeader() bool {
...
-
-	hdr, err := sr.BlockProcessor().CreateBlockHeader(
-		sr.BlockBody,
-		uint64(sr.Rounder().Index()),
-		haveTimeInCurrentSubround)
-	if err != nil {
-		return nil, err
-	}
-
+func (sr *SubroundBlock) createInitialHeaderData() (data.HeaderHandler, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/spos/commonSubround/subroundBlock.go]: done

File: integrationTests/testInitializer.go [hunk 4]
```diff
@@ -288,7 +288,18 @@ func CreateGenesisMetaBlock() *dataBlock.MetaBlock {
 func CreateIeleVMAndBlockchainHook(
 	accnts state.AccountsAdapter,
 ) (vmcommon.VMExecutionHandler, *hooks.VMAccountsDB) {
-	blockChainHook, _ := hooks.NewVMAccountsDB(accnts, TestAddressConverter)
+
+	args := hooks.ArgBlockChainHook{
+		Accounts:         accnts,
+		AddrConv:         TestAddressConverter,
+		StorageService:   &mock.ChainStorerMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testInitializer.go]: this test does not use system smart contracts running on metachain

File: integrationTests/testSyncNode.go [hunk 5]
```diff
@@ -92,6 +93,7 @@ func (tpn *TestProcessorNode) initBlockProcessorWithSync() {
 		StartHeaders:          tpn.GenesisBlocks,
 		RequestHandler:        tpn.RequestHandler,
 		Core:                  nil,
+		BlockChainHook:        &mock.BlockChainHookHandlerMock{},
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testSyncNode.go]: this test does not use system smart contracts running on metachain

File: process/block/metablock.go (hunk 1 above)
Comment (sasurobert) [reply process/block/metablock.go]: it is check on baseProcessor checkProcessorNilParameters

File: process/block/shardblock.go [hunk 6]
```diff
@@ -66,6 +66,7 @@ func NewShardProcessor(arguments ArgShardProcessor) (*shardProcessor, error) {
 		uint64Converter:               arguments.Uint64Converter,
 		onRequestHeaderHandlerByNonce: arguments.RequestHandler.RequestHeaderByNonce,
 		appStatusHandler:              statusHandler.NewNilStatusHandler(),
+		blockChainHook:                arguments.BlockChainHook,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/shardblock.go]: it is check on baseProcessor checkProcessorNilParameters

File: process/factory/shard/vmContainerFactory.go [hunk 7]
```diff
@@ -11,35 +10,24 @@ import (
...
-	accounts         state.AccountsAdapter
-	addressConverter state.AddressConverter
-	vmAccountsDB     *hooks.VMAccountsDB
-	cryptoHook       vmcommon.CryptoHook
+	vmAccountsDB *hooks.VMAccountsDB
+	cryptoHook   vmcommon.CryptoHook
 }
 
 // NewVMContainerFactory is responsible for creating a new virtual machine factory object
 func NewVMContainerFactory(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/shard/vmContainerFactory.go]: vmcontainer factory for shards creates other Virtual Machine.

File: process/smartContract/hooks/vmAccountsDB.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: done

File: process/smartContract/hooks/vmAccountsDB.go [hunk 8]
```diff
@@ -141,9 +181,106 @@ func (vadb *VMAccountsDB) GetCode(address []byte) ([]byte, error) {
...
+
+// CurrentRandomSeed returns the random seed from the current header
+func (vadb *VMAccountsDB) CurrentRandomSeed() []byte {
+	return vadb.currentHdr.GetRandSeed()
+}
+
+// CurrentEpoch returns the current epoch
+func (vadb *VMAccountsDB) CurrentEpoch() uint32 {
+	return vadb.currentHdr.GetEpoch()
 }
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: done

File: process/factory/metachain/vmContainerFactory_test.go [hunk 9]
```diff
@@ -1,44 +1,37 @@
 package metachain
 
 import (
+	"github.com/ElrondNetwork/elrond-go/data/state"
```
Comment (raduchis) [inline process/factory/metachain/vmContainerFactory_test.go]: sort imports

File: process/interface.go [hunk 10]
```diff
@@ -168,9 +167,9 @@ type BlockProcessor interface {
 	ProcessBlock(blockChain data.ChainHandler, header data.HeaderHandler, body data.BodyHandler, haveTime func() time.Duration) error
 	CommitBlock(blockChain data.ChainHandler, header data.HeaderHandler, body data.BodyHandler) error
 	RevertAccountState()
-	CreateBlockBody(round uint64, haveTime func() bool) (data.BodyHandler, error)
+	CreateBlockBody(initialHdrData data.HeaderHandler, haveTime func() bool) (data.BodyHandler, error)
 	RestoreBlockIntoPools(header data.HeaderHandler, body data.BodyHandler) error
-	CreateBlockHeader(body data.BodyHandler, round uint64, haveTime func() bool) (data.HeaderHandler, error)
+	ApplyBodyToHeader(hdr data.HeaderHandler, body data.BodyHandler) error
```
Comment (raduchis) [inline process/interface.go]: I would call it SetHeaderBody

File: process/smartContract/hooks/argHooks.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/data/typeConverters"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+// ArgBlockChainHook represents the arguments structure for the blockchain hook
+type ArgBlockChainHook struct {
```
Comment (raduchis) [inline process/smartContract/hooks/argHooks.go]: Call it maybe BlockChainHookArg

File: process/smartContract/hooks/vmAccountsDB.go [hunk 12]
```diff
@@ -2,45 +2,88 @@ package hooks
 
 import (
 	"encoding/binary"
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: sort imports

File: process/smartContract/hooks/vmAccountsDB.go [hunk 13]
```diff
@@ -2,45 +2,88 @@ package hooks
...
 )
 
 // VMAccountsDB is a wrapper over AccountsAdapter that satisfy vmcommon.BlockchainHook interface
 type VMAccountsDB struct {
-	accounts state.AccountsAdapter
-	addrConv state.AddressConverter
+	accounts         state.AccountsAdapter
+	addrConv         state.AddressConverter
+	storageService   dataRetriever.StorageService
+	blockChain       data.ChainHandler
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: Why does a VMAccountsDB need a blockchain and shardCoordinator?

Is VMAccountsDB more than a DB for the VM? Maybe change the name if it is more.

File: process/smartContract/hooks/vmAccountsDB.go [hunk 14]
```diff
@@ -141,9 +184,134 @@ func (vadb *VMAccountsDB) GetCode(address []byte) ([]byte, error) {
...
+
+	if requestedNonce.Cmp(big.NewInt(0)) < 0 {
+		return nil, process.ErrInvalidNonceRequest
+	}
+
+	if offset.Cmp(big.NewInt(0)) == 0 {
+		return vadb.blockChain.GetCurrentBlockHeaderHash(), nil
+	}
+
+	if vadb.shardCoordinator.SelfId() == sharding.MetachainShardId {
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: the process inside is of type MetaProcessor or ShardProcessor, could we simplify this by calling:

GetHeaderFromStorage(
requestedNonce.Uint64(),
		vadb.shardCoordinator.SelfId(),
		vadb.storageService,
		vadb.uint64Converter,
		vadb.marshalizer,
)

And behind call the different methods?

File: process/smartContract/hooks/vmAccountsDB.go [hunk 15]
```diff
@@ -265,3 +433,10 @@ func (vadb *VMAccountsDB) IsInterfaceNil() bool {
 	}
 	return false
 }
+
+// SetCurrentHeader sets current header to be used by smart contracts
+func (vadb *VMAccountsDB) SetCurrentHeader(hdr data.HeaderHandler) {
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: SC should not be aware of a block or header. A block or header contains SC.

File: process/smartContract/hooks/vmAccountsDB_test.go [hunk 16]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/process"
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB_test.go]: Fix imports

File: process/smartContract/process.go [hunk 17]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (raduchis) [inline process/smartContract/process.go]: fix imports

File: integrationTests/testSyncNode.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testSyncNode.go]: it is blockchainHook. not the effective chain

File: cmd/node/factory/structs.go [hunk 18]
```diff
@@ -7,6 +7,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/process/smartContract/hooks"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done

File: consensus/spos/commonSubround/subroundBlock.go [hunk 19]
```diff
@@ -2,6 +2,7 @@ package commonSubround
 
 import (
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/data/block"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/spos/commonSubround/subroundBlock.go]: done

File: consensus/spos/commonSubround/subroundBlock_test.go [hunk 20]
```diff
@@ -652,8 +652,11 @@ func TestSubroundBlock_CreateHeaderNilCurrentHeader(t *testing.T) {
 	}
 	container := mock.InitConsensusCore()
 	sr := *initSubroundBlock(blockChain, container)
-	sr.BlockChain().SetCurrentBlockHeader(nil)
+	_ = sr.BlockChain().SetCurrentBlockHeader(nil)
 	header, _ := sr.CreateHeader()
+	_ = sr.SendBlockBody(header)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/spos/commonSubround/subroundBlock_test.go]: this is the order in which subroundBlock calls the functions.

File: integrationTests/testProcessorNode.go [hunk 21]
```diff
@@ -677,8 +673,21 @@ func (tpn *TestProcessorNode) ProposeBlock(round uint64, nonce uint64) (data.Bod
 	buff, _ := TestMarshalizer.Marshal(currHdr)
 	blockHeader.SetPrevHash(TestHasher.Compute(string(buff)))
 	blockHeader.SetPrevRandSeed(currHdr.GetRandSeed())
+	sig, _ := TestMultiSig.AggregateSigs(nil)
+	blockHeader.SetSignature(sig)
 	blockHeader.SetRandSeed(sig)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testProcessorNode.go]: this is a testprocessornode, without that verification.

File: process/block/shardblock.go [hunk 22]
```diff
@@ -490,12 +493,14 @@ func (sp *shardProcessor) restoreMetaBlockIntoPool(miniBlockHashes map[string]ui
 
 // CreateBlockBody creates a a list of miniblocks by filling them with transactions out of the transactions pools
 // as long as the transactions limit for the block has not been reached and there is still time to add transactions
-func (sp *shardProcessor) CreateBlockBody(round uint64, haveTime func() bool) (data.BodyHandler, error) {
-	log.Debug(fmt.Sprintf("started creating block body in round %d\n", round))
+func (sp *shardProcessor) CreateBlockBody(initialHdrData data.HeaderHandler, haveTime func() bool) (data.BodyHandler, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/shardblock.go]: No. You need the initialHeader data for smart contracts which are using current nonce, randomness, epoch.

File: process/block/metablock.go [hunk 23]
```diff
@@ -1286,62 +1289,60 @@ func (mp *metaProcessor) createPeerInfo() ([]block.PeerData, error) {
 	return peerInfo, nil
 }
 
-// CreateBlockHeader creates a miniblock header list given a block body
-func (mp *metaProcessor) CreateBlockHeader(bodyHandler data.BodyHandler, round uint64, haveTime func() bool) (data.HeaderHandler, error) {
-	log.Debug(fmt.Sprintf("started creating block header in round %d\n", round))
+// ApplyBodyToHeader creates a miniblock header list given a block body
+func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler data.BodyHandler) error {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: You create half of the header in consensus, as there you have the data. Nonce, Round, Epoch, Signature, Randomness. Than add the data from body to it.

File: integrationTests/testProcessorNode.go [hunk 24]
```diff
@@ -653,22 +655,16 @@ func (tpn *TestProcessorNode) ProposeBlock(round uint64, nonce uint64) (data.Bod
...
-	if err != nil {
-		fmt.Println(err.Error())
-		return nil, nil, nil
-	}
-	blockHeader, err := tpn.BlockProcessor.CreateBlockHeader(blockBody, round, haveTime)
-	if err != nil {
-		fmt.Println(err.Error())
-		return nil, nil, nil
+	var blockHeader data.HeaderHandler
+	if tpn.ShardCoordinator.SelfId() == sharding.MetachainShardId {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testProcessorNode.go]: yes. we need that. done.

File: process/block/metablock.go [hunk 25]
```diff
@@ -463,12 +465,13 @@ func (mp *metaProcessor) RestoreBlockIntoPools(headerHandler data.HeaderHandler,
 }
 
 // CreateBlockBody creates block body of metachain
-func (mp *metaProcessor) CreateBlockBody(round uint64, haveTime func() bool) (data.BodyHandler, error) {
-	log.Debug(fmt.Sprintf("started creating block body in round %d\n", round))
+func (mp *metaProcessor) CreateBlockBody(initialHdrData data.HeaderHandler, haveTime func() bool) (data.BodyHandler, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: You create half of the header in consensus, as there you have the data. Nonce, Round, Epoch, Signature, Randomness. Than add the data from body to it.

File: process/smartContract/hooks/argHooks.go (hunk 11 above)
Comment (sasurobert) [reply process/smartContract/hooks/argHooks.go]: I followed the arguments naming convention already in use.

File: process/smartContract/hooks/vmAccountsDB.go (hunk 15 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: needs data from current round, which exists in header, and will be later signed by consensus

File: process/smartContract/hooks/vmAccountsDB_test.go (hunk 16 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB_test.go]: done

File: process/smartContract/process.go (hunk 17 above)
Comment (sasurobert) [reply process/smartContract/process.go]: done

File: process/smartContract/hooks/vmAccountsDB.go (hunk 12 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: done.

File: consensus/spos/commonSubround/subroundBlock.go [hunk 26]
```diff
@@ -187,21 +194,7 @@ func (sr *SubroundBlock) sendBlockHeader() bool {
...
-
-	hdr, err := sr.BlockProcessor().CreateBlockHeader(
-		sr.BlockBody,
-		uint64(sr.Rounder().Index()),
-		haveTimeInCurrentSubround)
-	if err != nil {
-		return nil, err
-	}
-
+	hdr := &block.Header{}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/spos/commonSubround/subroundBlock.go]: good catch.

File: consensus/spos/commonSubround/subroundBlock.go [hunk 27]
```diff
@@ -87,12 +86,30 @@ func (sr *SubroundBlock) doBlockJob() bool {
 		return false
 	}
 
-	if !sr.sendBlockBody() ||
-		!sr.sendBlockHeader() {
+	hdr, err := sr.createHeader()
```
Comment (iulianpascalau) [inline consensus/spos/commonSubround/subroundBlock.go]: :+1:

File: consensus/mock/blockProcessorMock.go [hunk 28]
```diff
@@ -16,11 +16,16 @@ type BlockProcessorMock struct {
 	CreateBlockCalled                func(initialHdrData data.HeaderHandler, haveTime func() bool) (data.BodyHandler, error)
 	RestoreBlockIntoPoolsCalled      func(header data.HeaderHandler, body data.BodyHandler) error
 	SetOnRequestTransactionCalled    func(f func(destShardID uint32, txHash []byte))
-	CreateBlockHeaderCalled          func(body data.BodyHandler, round uint64, haveTime func() bool) (data.HeaderHandler, error)
+	ApplyBodyToHeaderCalled          func(hdr data.HeaderHandler, body data.BodyHandler) error
 	MarshalizedDataToBroadcastCalled func(header data.HeaderHandler, body data.BodyHandler) (map[uint32][]byte, map[string][][]byte, error)
 	DecodeBlockBodyCalled            func(dta []byte) data.BodyHandler
 	DecodeBlockHeaderCalled          func(dta []byte) data.HeaderHandler
 	AddLastNotarizedHdrCalled        func(shardId uint32, processedHdr data.HeaderHandler)
+	CreateNewHeaderCalled            func() data.HeaderHandler
```
Comment (iulianpascalau) [inline consensus/mock/blockProcessorMock.go]: :+1:

File: process/block/shardblock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/block/shardblock.go]: Ok, arguments.Core and arguments.StartHeaders are not checked for nil

File: integrationTests/testInitializer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/testInitializer.go]: this test does not use system smart contracts running on metachain. those mock are only for the blockchain hook.

File: process/block/shardblock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/shardblock.go]: startHeaders will faild when setLastNotarizedHeadersSlice is called on the constructor of shardprocessor and metaprocessor as well.

File: consensus/spos/commonSubround/subroundBlock.go [hunk 29]
```diff
@@ -86,12 +87,18 @@ func (sr *SubroundBlock) doBlockJob() bool {
...
+	if err != nil {
+		log.Error(err.Error())
+		return false
+	}
+
+	if !sr.sendBlockBody(hdr) ||
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/spos/commonSubround/subroundBlock.go]: refactored

File: process/smartContract/hooks/vmAccountsDB.go (hunk 13 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: changed

File: process/block/shardblock.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/shardblock.go]: Core can be nil, as it is for elastic search / termui / etc.

File: integrationTests/testInitializer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/testInitializer.go]: Will be corrected when the new system VM integration tests will be written

[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/testInitializer.go]: ok

File: integrationTests/testSyncNode.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/testSyncNode.go]: ok

File: consensus/spos/commonSubround/subroundBlock_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply consensus/spos/commonSubround/subroundBlock_test.go]: the refactoring fixed the problem

File: integrationTests/testProcessorNode.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/testProcessorNode.go]: o...k...

File: process/smartContract/hooks/argHooks.go (hunk 11 above)
Comment (raduchis) [reply process/smartContract/hooks/argHooks.go]: ok... should be called with Arg at the end in the future

File: process/smartContract/hooks/vmAccountsDB.go (hunk 15 above)
Comment (raduchis) [reply process/smartContract/hooks/vmAccountsDB.go]: o...k....

File: process/smartContract/hooks/vmAccountsDB.go [hunk 30]
```diff
@@ -5,45 +5,88 @@ import (
 	"math/big"
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: Change also the fileName please: vmAAccountsDB to blockchainhookImpl

File: process/smartContract/hooks/vmAccountsDB.go [hunk 31]
```diff
@@ -141,26 +184,138 @@ func (vadb *VMAccountsDB) GetCode(address []byte) ([]byte, error) {
...
+
+	if requestedNonce.Cmp(big.NewInt(0)) < 0 {
+		return nil, process.ErrInvalidNonceRequest
+	}
+
+	if offset.Cmp(big.NewInt(0)) == 0 {
+		return bh.blockChain.GetCurrentBlockHeaderHash(), nil
+	}
+
+	_, hash, err := process.GetShardHeaderFromStorageWithNonce(
```
Comment (raduchis) [inline process/smartContract/hooks/vmAccountsDB.go]: Is it ok? Now you call just the GetShardHeaderWithNonce and not any more GetMetaHeaderWithNonce?

File: process/smartContract/hooks/vmAccountsDB.go (hunk 30 above)
Comment (sasurobert) [reply process/smartContract/hooks/vmAccountsDB.go]: done

File: process/common.go [hunk 32]
```diff
@@ -276,6 +276,21 @@ func GetMetaHeaderFromPoolWithNonce(
...
+// GetHeaderFromStorageWithNonce method returns a block header from storage with a given nonce and shardId
+func GetHeaderFromStorageWithNonce(
+	nonce uint64,
+	shardId uint32,
+	storageService dataRetriever.StorageService,
+	uint64Converter typeConverters.Uint64ByteSliceConverter,
+	marshalizer marshal.Marshalizer,
+) (data.HeaderHandler, []byte, error) {
+
+	if shardId == sharding.MetachainShardId {
```
Comment (iulianpascalau) [inline process/common.go]: This doesn't look right. I know how it is implemented but common.go contains a lot of almost duplicate helper functions. We should have had a common interface of these common funcs with 2 implementations. Might consider this in the next process refactor wave

**Pull request merged**